File: swim-js/swim-runtime/swim-host/swim-client/src/main/downlink/DownlinkReifier.ts
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Item, Field, Value, Record} from "@swim/structure";
import {RecordModel, Reifier} from "@swim/dataflow";
import {DownlinkStreamlet} from "./DownlinkStreamlet";
import type {WarpRef} from "../ref/WarpRef";

/** @internal */
export class DownlinkReifier extends Reifier {
  constructor(warp: WarpRef | null = null) {
    super();
    this.warp = warp;
  }

  warp: WarpRef | null;

  override reify(item: Item): Item {
    if (item instanceof Field) {
      return this.reifyField(item);
    } else {
      return this.reifyValue(item);
    }
  }

  /** @internal */
  reifyField(field: Field): Field {
    const oldValue = field.value;
    const newValue = this.reifyValue(oldValue);
    if (oldValue !== newValue) {
      return field.updatedValue(newValue);
    } else {
      return field;
    }
  }

  /** @internal */
  reifyValue(value: Value): Value {
    if (value instanceof RecordModel) {
      return this.reifyModel(value);
    } else {
      return value;
    }
  }

  /** @internal */
  reifyModel(model: RecordModel): Record {
    if (model.tag === "link") {
      const streamlet = new DownlinkStreamlet(this.warp, model);
      streamlet.compile();
      return streamlet;
    }
    return model;
  }
}

File: swim-java/swim-runtime/swim-host/swim.api/src/main/java/swim/api/downlink/DownlinkReifier.java
// Copyright 2015-2023 Nstream, inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.api.downlink;

import swim.api.ref.WarpRef;
import swim.dataflow.RecordModel;
import swim.dataflow.Reifier;
import swim.structure.Field;
import swim.structure.Item;
import swim.structure.Record;
import swim.structure.Value;

final class DownlinkReifier extends Reifier {

  final WarpRef warp;

  DownlinkReifier(WarpRef warp) {
    this.warp = warp;
  }

  @Override
  public Item reify(Item item) {
    if (item instanceof Field) {
      return this.reifyField((Field) item);
    } else {
      return this.reifyValue((Value) item);
    }
  }

  public Field reifyField(Field field) {
    final Value oldValue = field.value();
    final Value newValue = this.reifyValue(oldValue);
    if (oldValue != newValue) {
      return field.updatedValue(newValue);
    } else {
      return field;
    }
  }

  public Value reifyValue(Value value) {
    if (value instanceof RecordModel) {
      return this.reifyModel((RecordModel) value);
    } else {
      return value;
    }
  }

  public Record reifyModel(RecordModel model) {
    if ("link".equals(model.tag())) {
      final DownlinkStreamlet streamlet = new DownlinkStreamlet(this.warp, model);
      streamlet.compile();
      return streamlet;
    }
    return model;
  }

}
